### PR TITLE
[11.0][ADD] hr_timesheet_sheet_no_negative

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Split of https://github.com/OCA/hr-timesheet/pull/188.

Waiting the merge:
- [x] https://github.com/OCA/hr-timesheet/pull/207.

HR Timesheet Sheet No Negative
==============================

Negative hours are not admitted.

### CODE REVIEW

File: hr_timesheet_sheet_no_negative/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import models
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet_no_negative/__init__.py:3]: License missing

File: hr_timesheet_sheet/models/account_analytic_line.py [hunk 2]
```diff
@@ -85,6 +85,10 @@ def _check_state(self):
                     _('You cannot modify an entry in a confirmed '
                       'timesheet sheet.'))
 
+    def _negative_unit_amount(self):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/account_analytic_line.py]: I think it's not the best way to workaround, by default new modules should not break already-existing tests of other modules? @pedrobaeza yet this is not my call

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 3]
```diff
@@ -567,15 +567,25 @@ class SheetLine(models.TransientModel):
...
+        return True
+
     @api.onchange('unit_amount')
     def onchange_unit_amount(self):
         """This method is called when filling a cell of the matrix.
         It checks if there exists timesheets associated  to that cell.
         If yes, it does several comparisons to see if the unit_amount of
         the timesheets should be updated accordingly."""
         self.ensure_one()
-        if self.unit_amount < 0.0:
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Better extract it to "normalize unit amount" something

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 4]
```diff
@@ -644,7 +655,7 @@ def _create_timesheet(self, amount):
     @api.model
     def _diff_amount_timesheets(self, diff_amount, timesheets):
         for timesheet in timesheets:
-            if timesheet.unit_amount + diff_amount >= 0.0:
+            if self._is_positive(timesheet.unit_amount + diff_amount, True):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `is_valid_amount`?

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet_no_negative/__init__.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/__init__.py:3]: See https://github.com/OCA/maintainer-tools/commit/2ffdff1df7675fd6cab14e39e4ff5c6d46b9edb1#diff-2dc8828130dfbafb770bf7e7c4ae7557.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: ?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 4 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I prefer `_is_positive`

File: hr_timesheet_sheet_no_negative/__init__.py (hunk 1 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet_no_negative/__init__.py:3]: Not copyright, license https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 3 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `self.unit_amount = self._normalize_unit_amount(self.unit_amount)` ?

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: or `adjust_unit_amount`, as extension point

File: hr_timesheet_sheet_attendance/models/hr_attendance.py [hunk 5]
```diff
@@ -57,6 +57,7 @@ def _compute_sheet_id(self):
     sheet_id = fields.Many2one(
         comodel_name='hr_timesheet.sheet',
         compute="_compute_sheet_id",
+        compute_sudo=True,
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet_attendance/models/hr_attendance.py]: Is this related to the PR?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `adjust_unit_amount` is an empty method?

File: hr_timesheet_sheet_no_negative/__init__.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/__init__.py:3]: ah, sorry

File: hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py [hunk 6]
```diff
@@ -321,8 +321,12 @@ def test_5(self):
                 {'unit_amount': -1.0}, update=True))
         line.onchange_unit_amount()
         sheet._onchange_dates_or_timesheets()
-        self.assertEqual(len(sheet.line_ids), 0)
-        self.assertEqual(len(sheet.timesheet_ids), 0)
+        if line._negative_unit_amount():
+            self.assertEqual(len(sheet.line_ids), 7)
+            self.assertEqual(len(sheet.timesheet_ids), 1)
+        else:
+            self.assertEqual(len(sheet.line_ids), 0)
```
Comment (astirpe) [inline hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py]: Why do you need this `if _negative_unit_amount(): - else:` block in `hr_timesheet_sheet`?
This block and the similar others should be moved to the tests of the other module.

This way you can also avoid to define an empty method `_negative_unit_amount()` in module `hr_timesheet_sheet`.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 3 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Yeah, good idea!

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 4 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I'm just trying to avoid specific positive/negative wording, and add generic extension point  :)

**Review by astirpe — CHANGES_REQUESTED** [review]

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py (hunk 6 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py]: It's because I twas too lazy to copy the tests haha

File: hr_timesheet_sheet_attendance/models/hr_attendance.py (hunk 5 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_attendance/models/hr_attendance.py]: No, I was just trying to solve the hr_timesheet_sheet_attendance travis error

Comment (alexey-pelykh) [reply hr_timesheet_sheet_attendance/models/hr_attendance.py]: Got it, I know it takes time, yet maybe let's move it in another PR? :)))

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_attendance/models/hr_attendance.py]: Why? I don't want to create 1000 PRs. This travis error is unrelated to this PR, true, but it's also true that this PR will not be merged until this travis error is solved. Thus, I will still try to solve this error in the same PR.

Comment (alexey-pelykh) [reply hr_timesheet_sheet_attendance/models/hr_attendance.py]: Valid point, I have no idea how this should be handled. @pedrobaeza share your the knowledge please :)

**Comment by MiquelRForgeFlow** [comment]
@astirpe @alexey-pelykh  ready to be reviewed :)

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 7]
```diff
@@ -568,15 +568,26 @@ class SheetLine(models.TransientModel):
         default=0,
     )
 
+    def _is_positive(self, value, equal=False):
+        if value > 0.0:
+            return True
+        elif value == 0.0 and equal:
```
Comment (leemannd) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]:
```suggestion
        elif float_is_zero(value) and equal:
```

Comment (leemannd) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: and as import:
```
from odoo.tools import float_is_zero
```
It can also be used, if needed with the digits precition or rounding if needed

**Review by leemannd — COMMENTED** [review]
Please consider using `float_is_zero`

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 7 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: But which precision? `float_is_zero` requires a precision number

**Comment by astirpe** [comment]
Isn't commit [[FIX] hr_timesheet_sheet_attendance: AccessError](https://github.com/OCA/hr-timesheet/pull/201/commits/609aac49310920ee314002999798833f128b12fc) unrelated to this PR?

**Comment by MiquelRForgeFlow** [comment]
Yeah, totally unrelated, but without it, the PR is red like your other PR. And you know, nobody approves red PRs.

**Comment by MiquelRForgeFlow** [comment]
@astirpe @alexey-pelykh could you update review, please?

File: .travis.yml [hunk 8]
```diff
@@ -20,8 +20,10 @@ env:
 
   matrix:
   - LINT_CHECK="1"
-  - TESTS="1" ODOO_REPO="OCA/OCB" MAKEPOT="1"
-  - TESTS="1" ODOO_REPO="odoo/odoo"
+  - TESTS="1" ODOO_REPO="OCA/OCB" MAKEPOT="1" EXCLUDE=hr_timesheet_task_required
```
Comment (alexey-pelykh) [inline .travis.yml]: @mreficent why this is needed?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 9]
```diff
@@ -582,7 +591,7 @@ def onchange_unit_amount(self):
                 new_ts = new_ts.merge_timesheets()
             unit_amount = new_ts.unit_amount + diff_amount
             new_ts.write({'unit_amount': unit_amount})
-        else:
+        elif self._adjust_amount(diff_amount):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Not sure this is correct - you're ignoring the return value because?

File: hr_timesheet_sheet_no_negative/models/__init__.py [hunk 10]
```diff
@@ -0,0 +1,4 @@
+from . import account_analytic_line
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet_no_negative/models/__init__.py:3]: License lost

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+        <field name="name">res.config.settings.view.form.inherit.hr.timesheet.no.negative</field>
+        <field name="model">res.config.settings</field>
+        <field name="inherit_id" ref="hr_timesheet_sheet.res_config_settings_view_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//div[@name='timesheet_week_start']" position="after">
+                <div class="col-xs-12 col-md-6 o_setting_box">
+                    <div class="o_setting_right_pane">
+                        <label for="timesheet_negative_unit_amount"/>
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml:11]: Shoudn't this have an icon about being company-specific?

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 12]
```diff
@@ -548,25 +548,34 @@ class SheetLine(models.TransientModel):
...
 
-    @api.onchange('unit_amount')
-    def onchange_unit_amount(self):
-        """This method is called when filling a cell of the matrix.
-        It checks if there exists timesheets associated  to that cell.
-        If yes, it does several comparisons to see if the unit_amount of
-        the timesheets should be updated accordingly."""
-        self.ensure_one()
+    def _adjust_amount(self, value):
+        """For extensions"""
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Can't you just use `_line_to_timesheet()` for extensions? I don't see the need of adding this method.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 13]
```diff
@@ -548,25 +548,34 @@ class SheetLine(models.TransientModel):
...
+        the timesheets should be updated accordingly."""
+        self.ensure_one()
+        sheet = self._get_sheet()
         if not sheet:
             return {'warning': {
                 'title': _("Warning"),
                 'message': _("Save the Timesheet Sheet first.")
             }}
+        elif not self.sheet_id:
+            self.sheet_id = sheet
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I thought this is was not necessary, did you find a case where this is needed?

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    def onchange_unit_amount(self):
+        self.ensure_one()
+        sheet = self._get_sheet()
+        if not sheet:
+            return {'warning': {
+                'title': _("Warning"),
+                'message': _("Save the Timesheet Sheet first.")
+            }}
+        elif not self.sheet_id:
+            self.sheet_id = sheet
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: I see 8 lines of code duplicated from the main module. There's no better solution? :thinking:

File: .travis.yml (hunk 8 above)
Comment (MiquelRForgeFlow) [reply .travis.yml]: Because `hr_timesheet_task_required` is installed before `hr_timesheet_sheet_no_negative`, and thus in the `hr_timesheet_sheet_no_negative` tests the task is required, and I don't want to put task in the tests.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 9 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: If you are forbidding negative values, then you don't want to create timesheets with negative value. Thus, in this `elif` we are checking if `diff_amount` is negative or not.

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 11 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml:11]: Can you link me an example, please?

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = 'account.analytic.line'
+
+    def _negative_unit_amount(self):
+        if not self.company_id:
+            user_company = self.env.user.company_id
+            company_id = self._context.get('company_id') or user_company.id
+            company = self.env['res.company'].browse(company_id)
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/account_analytic_line.py:12]: `user_company` is already a record so, in case no `company_id` in context, you are converting it to an `id` and then back to a record. I think this block could be written in a smarter way.
Also `company_id` is a related field and it's required in `account.analytic.account`, so I think this block of code is not necessary at all.

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 11 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml:11]: https://github.com/OCA/hr-timesheet/blob/12.0/hr_timesheet_role/views/res_config_settings.xml#L21

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml [hunk 16]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
+
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+        <field name="name">res.config.settings.view.form.inherit.hr.timesheet.no.negative</field>
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: You don't need this long (and difficult to read) line. Better let the name be generated automatically by the system IMO.

File: .travis.yml (hunk 8 above)
Comment (alexey-pelykh) [reply .travis.yml]: But this would effectively double the CI time

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 9 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Not sure I'm following, by default `_adjust_amount` returns unchanged value, right? So, elif -1: will still execute, thus I'm not sure I follow

File: hr_timesheet_sheet_no_negative/README.rst [hunk 17]
```diff
@@ -0,0 +1,57 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+==============================
+HR Timesheet Sheet No Negative
+==============================
+
+This module disable negative hours.
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/README.rst]: I think you should clarify here that it's a functionality regarding the filling of timesheet sheet lines.

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 11 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml:11]: thanks

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+            company = self.company_id
+        allow_negative = company.timesheet_negative_unit_amount
+        return allow_negative
+
+    @api.onchange('unit_amount')
+    def onchange_unit_amount(self):
+        self.ensure_one()
+        if not self._negative_unit_amount() and self.project_id \
+                and self.unit_amount < 0.0:
+            self.write({'unit_amount': 0.0})
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/account_analytic_line.py:21]: Here you prevent the user to enter negative amounts on an analytic line, regardless whether the line belongs to a timesheet sheet or not, but the module is named `hr_timesheet_sheet_no_negative`. :thinking: Anyway, in general, I still have doubts about why this module would be needed.

**Review by astirpe — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 16 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: 😂 that's just tastes

File: .travis.yml (hunk 8 above)
Comment (alexey-pelykh) [reply .travis.yml]: Maybe let's just add context flag to exclude that check?

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py (hunk 18 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/account_analytic_line.py:21]: Well, the `and self.project_id` assures only applies to timesheets, not to any analytic line.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 13 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Didn't checked. I just put it to be sure because I needed `self.sheet_id` in `_negative_unit_amount` method.

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py (hunk 14 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: Any suggestion how to do it, please?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 9 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Well, now this `_adjust_amount` has disappeared :)

File: .travis.yml (hunk 8 above)
Comment (MiquelRForgeFlow) [reply .travis.yml]: done the context flag

File: hr_timesheet_task_required/models/account_analytic_line.py [hunk 19]
```diff
@@ -14,6 +15,9 @@ class AccountAnalyticLine(models.Model):
         'task_id',
     )
     def _check_timesheet_task(self):
+        if config['test_enable'] and not self.env.context.get(
+                'test_hr_timesheet_task_required'):
```
Comment (alexey-pelykh) [inline hr_timesheet_task_required/models/account_analytic_line.py:19]: Shouldn't it be inverse? like `skip_hr_timesheet_task_required` that would by default be None/False?

Comment (alexey-pelykh) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: @mreficent ^

Comment (MiquelRForgeFlow) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: I just copied the fix you did for `hr_timesheet_employee_required` module.

Comment (alexey-pelykh) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: My case is different - `hr_timesheet_employee_required` affects all modules by being present in the system (has no setting), thus I'm explicitly telling that it's ok to run tests. `hr_timesheet_task_required` on contrary does not break anything by being present, as by default it's not requiring task

Comment (MiquelRForgeFlow) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: `hr_timesheet_task_required` by default is requiring task. The requirement is just the existence of the constrains.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 20]
```diff
@@ -584,7 +589,8 @@ def onchange_unit_amount(self):
             new_ts.write({'unit_amount': unit_amount})
         else:
             new_ts_values = self._line_to_timesheet(diff_amount)
-            self.env['account.analytic.line'].create(new_ts_values)
+            if new_ts_values:
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:589]: Why?

File: hr_timesheet_task_required/models/account_analytic_line.py (hunk 19 above)
Comment (alexey-pelykh) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: Ah, so sorry - that's for v11, for v12 it's https://github.com/OCA/hr-timesheet/blob/12.0/hr_timesheet_task_required/models/res_company.py#L10

Comment (alexey-pelykh) [reply hr_timesheet_task_required/models/account_analytic_line.py:19]: So yes, for v11 `hr_timesheet_task_required` your fix is valid

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 13 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: So: marked as resolved as a "just in case"? Not going to give it a try without those 2 lines of code?

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py (hunk 18 above)
Comment (astirpe) [reply hr_timesheet_sheet_no_negative/models/account_analytic_line.py:21]: I was mentioning timesheet sheets.

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+class SheetLine(models.TransientModel):
+    _inherit = 'hr_timesheet.sheet.line'
+
+    def _negative_unit_amount(self):
+        if not self.sheet_id.company_id:
+            if self._context.get('company_id'):
+                company_id = self._context.get('company_id')
+                company = self.env['res.company'].browse(company_id)
+            else:
+                company = self.env.user.company_id
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: Better: `self.employee_id.company_id`

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+        else:
+            company = self.sheet_id.company_id
+        allow_negative = company.timesheet_negative_unit_amount
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: Just return here

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py (hunk 14 above)
Comment (astirpe) [reply hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: At least would you check whether the last two lines are necessary?

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 16 above)
Comment (astirpe) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: I would prefer to see this monster line removed.

**Review by astirpe — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 20 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:589]: because now `self._line_to_timesheet` can return False in `hr_timesheet_sheet_no_negative` module.

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py (hunk 18 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/account_analytic_line.py:21]: No, you mentioned "analytic line" before. So I explained that the condition doesn't apply to all analytic lines, only applies to those that have a project, and that means, to those that its `unit_amount` means `hours`.

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 16 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: Is the generated automatically shorter?

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py (hunk 14 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: removed...

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 16 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: removed, but I left it commented to avoid anybody to fill it with another name

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py (hunk 21 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: done

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py (hunk 22 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: that is just cosmetic change but done

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 13 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: removed those 2 lines...

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 20 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:589]: You could return `{}` instead, so you can avoid this `if`.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:589]: No, because then `self.env['account.analytic.line'].create({})` would create a timesheet. Well, in fact, it will give an error because no required parameter passed in the `create`.

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py (hunk 18 above)
Comment (astirpe) [reply hr_timesheet_sheet_no_negative/models/account_analytic_line.py:21]: I see that you updated the readme, explaining that it also applies to the timesheets and not only to the timesheets lines.
:+1:

File: hr_timesheet_sheet_no_negative/models/account_analytic_line.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+        if self._context.get('company_id'):
+            company_id = self._context.get('company_id')
+            company = self.env['res.company'].browse(company_id)
+        else:
+            company = self.company_id
+        return company.timesheet_negative_unit_amount
+
+    @api.onchange('unit_amount')
+    def onchange_unit_amount(self):
+        self.ensure_one()
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/account_analytic_line.py]: You don't need this `ensure_one()` since it's an onchange.

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+                company = self.env['res.company'].browse(company_id)
+            else:
+                company = self.sheet_id.employee_id.company_id
+        else:
+            company = self.sheet_id.company_id
+        return company.timesheet_negative_unit_amount
+
+    @api.onchange('unit_amount')
+    def onchange_unit_amount(self):
+        self.ensure_one()
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: You don't need this `ensure_one()`. Could be removed in the main module as well, but anyway it's just a minor detail.

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+        sheet = self._get_sheet()
+        if not sheet:
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: `if not self._get_sheet():`

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+            return {'warning': {
+                'title': _("Warning"),
+                'message': _("Save the Timesheet Sheet first.")
+            }}
+        if not self._negative_unit_amount() and self.unit_amount < 0.0:
+            self.write({'unit_amount': 0.0})
+        super(SheetLine, self).onchange_unit_amount()
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py]: Since you are inheriting from an onchange, here you should return the result of the super, otherwise you could break some eventual other `onchange_unit_amount()` customization.

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml [hunk 27]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
+
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+        <!--<field name="name">res.config.settings.view.form.inherit.hr.timesheet.no.negative</field>-->
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: Just commented? :open_mouth:

**Review by astirpe — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml (hunk 27 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet_no_negative/views/res_config_settings_views.xml]: you really hate this name 😂😂
ok, removed

File: hr_timesheet_sheet_no_negative/models/res_company.py [hunk 28]
```diff
@@ -0,0 +1,12 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    timesheet_negative_unit_amount = fields.Boolean(
+        string="Negative Quantities",
+        default=False,
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet_no_negative/models/res_company.py:11]: That's already default, no need to set explicitly

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+            else:
+                company = self.sheet_id.employee_id.company_id
+        else:
+            company = self.sheet_id.company_id
+        return company.timesheet_negative_unit_amount
+
+    @api.onchange('unit_amount')
+    def onchange_unit_amount(self):
+        if not self._get_sheet():
+            return {'warning': {
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py:23]: What do you think about:

```
if not self._get_sheet():
    return super(SheetLine, self).onchange_unit_amount()
```

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+    def _negative_unit_amount(self):
+        if not self.sheet_id.company_id:
+            if self._context.get('company_id'):
+                company_id = self._context.get('company_id')
+                company = self.env['res.company'].browse(company_id)
+            else:
+                company = self.sheet_id.employee_id.company_id
+        else:
+            company = self.sheet_id.company_id
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py:17]: You can put this line at the beginning:

```
company = self.sheet_id.company_id
if not company:
    ...
```
so you save another line.

File: hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, models, _
+
+
+class SheetLine(models.TransientModel):
+    _inherit = 'hr_timesheet.sheet.line'
+
+    def _negative_unit_amount(self):
+        if not self.sheet_id.company_id:
+            if self._context.get('company_id'):
```
Comment (astirpe) [inline hr_timesheet_sheet_no_negative/models/hr_timesheet_sheet.py:11]: I read that `self._context` is deprecated, prefer to use `self.env.context` instead.

**Review by astirpe — CHANGES_REQUESTED** [review]

**Comment by MiquelRForgeFlow** [comment]
Not needed.
